### PR TITLE
Base Schema.org Implementation

### PR DESCRIPTION
**Pull request by shazmasiddiqui**
Pull Request for Issue #5

### CODE REVIEW

File: libraries/src/Schemaorg/SchemaorgPluginTrait.php [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+     * updates schema form
+     *
+     * @param   Form $form  Form to manipulate
+     * @param   $schemaType Schema Type to add
+     *
+     * @return  boolean
+     *
+     * @since   4.0.0
+     */
+    protected function updateSchemaForm($data)
```
Comment (ditsuke) [inline libraries/src/Schemaorg/SchemaorgPluginTrait.php]: What does "update" mean?

File: plugins/schemaorg/recipe/recipe.php [hunk 2]
```diff
@@ -56,11 +72,22 @@ public function onSchemaPrepareForm(Form $form)
...
     }
+
+    /**
+     *  Add a new option to the schema type in the article editing page
+     *
+     *  @param   Form  $form  The form to be altered.
+     *
+     *  @return  boolean
+     */
+    public function onSchemaBeforeSave(AbstractEvent $event)
```
Comment (ditsuke) [inline plugins/schemaorg/recipe/recipe.php]: You don't need to have special function names for listeners. Offer a default subscriber map to [`onSchedmaBeforeSave` => `saveSchema`] so the schema plugin can use it in directly in `getSubscribedEvents()`

File: libraries/src/Schemaorg/SchemaorgPluginTrait.php [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+     * Saves unfiltered JSON data of the form fields in database
+     *
+     * @param   Form $form  Form to manipulate
+     * @param   $schemaType Schema Type to add
+     *
+     * @return  boolean
+     *
+     * @since   4.0.0
+     */
+    protected function saveSchema(AbstractEvent $event, $form)
```
Comment (ditsuke) [inline libraries/src/Schemaorg/SchemaorgPluginTrait.php]: We should use a more descriptive name. An example: `storeSchemaToStandardLocation()` -- anything that tells the user our trait function saves the schema to the standard location.

**Review by ditsuke — CHANGES_REQUESTED** [review]

File: libraries/src/Schemaorg/SchemaorgPluginTrait.php (hunk 1 above)
Comment (shazmasiddiqui) [reply libraries/src/Schemaorg/SchemaorgPluginTrait.php]: This function will update the Schema form with prefilled data from database if article already has some data.

File: plugins/schemaorg/recipe/recipe.php (hunk 2 above)
Comment (shazmasiddiqui) [reply plugins/schemaorg/recipe/recipe.php]: Okay

File: libraries/src/Schemaorg/SchemaorgPluginTrait.php [hunk 4]
```diff
@@ -0,0 +1,285 @@
...
+     *
+     * @param   Form $form  Form to manipulate
+     * @param   $type Schema Type to add
+     * @param   $value Value of the type
+     *
+     * @return  boolean
+     *
+     * @since   4.0.0
+     */
+    protected function addSchemaType(Form $form, $type)
```
Comment (ditsuke) [inline libraries/src/Schemaorg/SchemaorgPluginTrait.php]: Add return type annotation

File: libraries/src/Schemaorg/SchemaorgPluginTrait.php [hunk 5]
```diff
@@ -0,0 +1,285 @@
...
+     * Saves unfiltered and filtered JSON data of the form fields in database
+     *
+     * @param   $event  EventInterface
+     * @param   $form   Name of the form
+     *
+     * @return  boolean
+     *
+     * @since   4.0.0
+     */
+    protected function saveSchema($event)
```
Comment (ditsuke) [inline libraries/src/Schemaorg/SchemaorgPluginTrait.php]: Add return type annotation

File: libraries/src/Schemaorg/SchemaorgPluginTrait.php [hunk 6]
```diff
@@ -0,0 +1,285 @@
...
+
+
+    /**
+     *  To add plugin specific functions
+     *
+     *  @param   Registry $schema Schema form
+     *
+     *  @return  boolean
+     */
+    protected function cleanupIndividualSchema(Registry $schema)
```
Comment (ditsuke) [inline libraries/src/Schemaorg/SchemaorgPluginTrait.php]: If this is a `@todo`, please mention it in a code comment.

File: libraries/src/Schemaorg/SchemaorgPluginTrait.php [hunk 7]
```diff
@@ -0,0 +1,285 @@
...
+
+    /**
+     *  To change hour and mins to duration ISO format
+     *
+     *  @param   Registry $schema Schema form
+     *  @param   Array $durationKeys Keys with duration fields
+     *
+     *  @return  boolean
+     */
+    protected function changeDurationFormat(Registry $schema, $durationKeys)
```
Comment (ditsuke) [inline libraries/src/Schemaorg/SchemaorgPluginTrait.php]: A function name like `normalizeDurationsToISO` could be more indicative of its function.

Comment (ditsuke) [inline libraries/src/Schemaorg/SchemaorgPluginTrait.php]: Type annotation for second param and return.

File: libraries/src/Schemaorg/SchemaorgPluginTrait.php [hunk 8]
```diff
@@ -0,0 +1,285 @@
...
+
+    /**
+     *  To create an array from repeatable text field data
+     *
+     *  @param   Registry $schema Schema form
+     *  @param   Array $durationKeys Keys with duration fields
+     *
+     *  @return  array
+     */
+    protected function convertToArray(Registry $schema, $repeatableFields)
```
Comment (ditsuke) [inline libraries/src/Schemaorg/SchemaorgPluginTrait.php]: Clarify why we need this function and document an example of the transformation it's performing in the phpdoc comment.

File: plugins/schemaorg/recipe/recipe.php [hunk 9]
```diff
@@ -56,11 +76,40 @@ public function onSchemaPrepareForm(Form $form)
...
     }
+
+    /**
+     *  Saves the schema to the database
+     *
+     *  @param   AbstractEvent $event
+     *
+     *  @return  boolean
+     */
+    public function onSchemaAfterSave(AbstractEvent $event)
```
Comment (ditsuke) [inline plugins/schemaorg/recipe/recipe.php:130]: You do not need the `onXXX` functions. Map the functions to events in `getSubscribedEvents()` instead.

**Review by ditsuke — CHANGES_REQUESTED** [review]

File: libraries/src/Schemaorg/SchemaorgPluginTrait.php (hunk 1 above)
Comment (ditsuke) [reply libraries/src/Schemaorg/SchemaorgPluginTrait.php]: Please document this in the method phpdoc

File: libraries/src/Schemaorg/SchemaorgPluginTrait.php [hunk 10]
```diff
@@ -0,0 +1,285 @@
...
+                }
+
+                if ($newDuration) {
+                    $schema->set($durationKey, $newDuration);
+                } else {
+                    $schema->remove($durationKey);
+                }
+            }
+        }
+        $schema->toArray();
```
Comment (ditsuke) [inline libraries/src/Schemaorg/SchemaorgPluginTrait.php]: Why are returning the schema as an array?

**Pull request closed without merging**